tests/ut/azure_iot_json_reader_ut.c

```diff
@@ -0,0 +1,582 @@
+                                               0 ),
+                      eAzureIoTErrorInvalidArgument );
+}
+
+static void testAzureIoTJSONReader_Init_Success( void ** ppvState )
+{
+    AzureIoTJSONReader_t xReader;
+    uint8_t * pucTestJSON = azureiotjsonreaderutTEST_JSON;
+
+    assert_int_equal( AzureIoTJSONReader_Init( &xReader,
+                                               pucTestJSON,
+                                               strlen( pucTestJSON ) ),
+                      eAzureIoTSuccess );
+}
+
+static void testAzureIoTJSONReader_NextToken_Failure( void ** ppvState )
+{
+    /* Fail init if JSON reader is NULL */
+    assert_int_equal( AzureIoTJSONReader_NextToken( NULL ),
+                      eAzureIoTErrorInvalidArgument );
+}
+
+static void testAzureIoTJSONReader_NextToken_Success( void ** ppvState )
+{
+    AzureIoTJSONReader_t xReader;
+    AzureIoTJSONTokenType_t xTokenType;
+
+    assert_int_equal( AzureIoTJSONReader_Init( &xReader, azureiotjsonreaderutTEST_JSON, strlen( azureiotjsonreaderutTEST_JSON ) ),
+                      eAzureIoTSuccess );
+
+    /*Begin object */
+    assert_int_equal( AzureIoTJSONReader_NextToken( &xReader ),
+                      eAzureIoTSuccess );
+    assert_int_equal( AzureIoTJSONReader_TokenType( &xReader, &xTokenType ), eAzureIoTSuccess );
+    assert_int_equal( xTokenType, eAzureIoTJSONTokenBEGIN_OBJECT );
+
+    /*Property name */
+    assert_int_equal( AzureIoTJSONReader_NextToken( &xReader ),
+                      eAzureIoTSuccess );
+
+    assert_int_equal( AzureIoTJSONReader_TokenType( &xReader, &xTokenType ), eAzureIoTSuccess );
+    assert_int_equal( xTokenType, eAzureIoTJSONTokenPROPERTY_NAME );
+
+    /*Property string */
+    assert_int_equal( AzureIoTJSONReader_NextToken( &xReader ),
+                      eAzureIoTSuccess );
+
+    assert_int_equal( AzureIoTJSONReader_TokenType( &xReader, &xTokenType ), eAzureIoTSuccess );
+    assert_int_equal( xTokenType, eAzureIoTJSONTokenSTRING );
+
+    /*Property name */
+    assert_int_equal( AzureIoTJSONReader_NextToken( &xReader ),
+                      eAzureIoTSuccess );
+
+    assert_int_equal( AzureIoTJSONReader_TokenType( &xReader, &xTokenType ), eAzureIoTSuccess );
+    assert_int_equal( xTokenType, eAzureIoTJSONTokenPROPERTY_NAME );
+
+    /*Property string */
+    assert_int_equal( AzureIoTJSONReader_NextToken( &xReader ),
+                      eAzureIoTSuccess );
+
+    assert_int_equal( AzureIoTJSONReader_TokenType( &xReader, &xTokenType ), eAzureIoTSuccess );
+    assert_int_equal( xTokenType, eAzureIoTJSONTokenSTRING );
+
+    /*End object */
+    assert_int_equal( AzureIoTJSONReader_NextToken( &xReader ),
+                      eAzureIoTSuccess );
+
+    assert_int_equal( AzureIoTJSONReader_TokenType( &xReader, &xTokenType ), eAzureIoTSuccess );
+    assert_int_equal( xTokenType, eAzureIoTJSONTokenEND_OBJECT );
+}
+
+static void testAzureIoTJSONReader_SkipChildren_Failure( void ** ppvState )
```

Comment:
Per offline discussion, I think our main concern with these tests is whether they're invoking the embedded API correctly and to Himanshu's point whether they're processing errors from embedded SDK correctly.  So could we (or should we?) throw in more mock based tests, where basically we make sure that if say `az_foo_bar` fails than the `AzureIoTJSON_FooBar()` will fail and with the expected error code?

Reply:
Yup definitely looking to add those

Reply:
One said part is PRECONDITION makes the caller responsible for error checking... so it is more like middleware layer to make sure we handle error for all the cases.

Reply:
As discussed tracking for now:
https://github.com/Azure/azure-iot-middleware-freertos/issues/136